### PR TITLE
[Converter] Add subplugins for python custom converter

### PR DESCRIPTION
**Pull request by gichan-jang**
Add subplugins for python custom converter

*Flexbuffer Python API is not officially released yet, disable the test if the version of flatbuffers is lower than 2.0.
(https://github.com/google/flatbuffers/issues/5306)

Signed-off-by: Gichan Jang <gichan2.jang@samsung.com>

Self evaluation:
1. Build test: [ *]Passed [ ]Failed [ ]Skipped
2. Run test: [ *]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: ext/nnstreamer/tensor_converter/tensor_converter_python3.cc [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+
+  PyObject *param = PyList_New (0);
+  PyObject *input_array = PyArray_SimpleNewFromData (
+      1, input_dims, NPY_UINT8, in_info.data);
+  PyList_Append (param, input_array);
+
+  Py_LOCK ();
+  if (!PyObject_HasAttrString (core_obj, (char *)"convert")) {
+    Py_ERRMSG ("Cannot find 'invoke'");
+    goto done;
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_converter/tensor_converter_python3.cc:243]: Please check the pairs of `Py_LOCK` and `PY_LOCK` in this function.

Comment (gichan-jang) [reply ext/nnstreamer/tensor_converter/tensor_converter_python3.cc:243]: Right. Updated.

**Comment by dongju-chae** [comment]
BTW, flatbuffers 1.13 is planned? I thought that flatbuffers 2.0 is released soon.

**Review by dongju-chae — APPROVED** [review]
LGTM

File: ext/nnstreamer/tensor_converter/tensor_converter_python3.cc [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+ * GStreamer / NNStreamer tensor_converter subplugin, "python3"
+ * Copyright (C) 2021 Gichan Jang <gichan2.jang@samsung.com>
+ */
+/**
+ * @file	tensor_converter_python3.c
+ * @date	May 03 2021
+ * @brief	NNStreamer tensor-converter subplugin, "python3",
+*         which converts to tensors using python.
+ * @see		https://github.com/nnstreamer/nnstreamer
+ * @author	Gicchan Jang <gichan2.jang@samsung.com>
```
Comment (kimjh12) [inline ext/nnstreamer/tensor_converter/tensor_converter_python3.cc:12]: Minor : your name typo

File: ext/nnstreamer/tensor_converter/tensor_converter_python3.cc [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+  PyErr_Clear ();
+
+  dlclose (handle);
+  g_mutex_clear (&py_mutex);
+}
+
+/**
+ * @brief parse the converting result to feed output tensors
+ * @param[result] Python object retunred by convert
+ * @param[config] config Structure for output tensors config
```
Comment (kimjh12) [inline ext/nnstreamer/tensor_converter/tensor_converter_python3.cc:172]: param description may be wrong?

File: ext/nnstreamer/tensor_converter/tensor_converter_python3.cc (hunk 2 above)
Comment (gichan-jang) [reply ext/nnstreamer/tensor_converter/tensor_converter_python3.cc]: Updated.

File: ext/nnstreamer/tensor_converter/tensor_converter_python3.cc (hunk 3 above)
Comment (gichan-jang) [reply ext/nnstreamer/tensor_converter/tensor_converter_python3.cc]: Updated. Thanks.

**Comment by gichan-jang** [comment]
> @gichan-jang BTW, flatbuffers 1.13 is planned? I thought that flatbuffers 2.0 is released soon.

You're right. They appear to be preparing for 2.0. (https://github.com/google/flatbuffers/issues/6353)
`flexbuffers.py` will include in the version.

File: gst/nnstreamer/include/nnstreamer_plugin_api_converter.h [hunk 4]
```diff
@@ -69,6 +69,19 @@ typedef struct _NNStreamerExternalConverter
    * @param[in] config The config of output tensor/tensors
    * @retval Return subplugin caps (if config is NULL, return default caps)
    */
+
+  int (*open) (gchar *script_path, void **priv_data);
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_converter.h:73]: constify script_path?

Comment (gichan-jang) [reply gst/nnstreamer/include/nnstreamer_plugin_api_converter.h:73]: Updated.

File: gst/nnstreamer/include/nnstreamer_plugin_api_converter.h [hunk 5]
```diff
@@ -69,6 +69,19 @@ typedef struct _NNStreamerExternalConverter
...
+  /**< tensor_converter will call this to open subplugin.
+   * @param[in] script_path script path of the subplugin.
+   * @param[in/out] private_data A subplugin may save its internal private data here. The subplugin is responsible for alloc/free of this pointer. Normally, open() allocates memory for private_data.
+   * @return 0 if ok. < 0 if error.
+   */
+
+  void (*close) (void **priv_data);
+  /**< tensor_converter will call this to close subplugin.
+   * @param[in] private_data frees private_data and set NULL.
+   */
```
[the remark this answers is not in the document]
Comment (gichan-jang) [reply gst/nnstreamer/include/nnstreamer_plugin_api_converter.h:83]: `PyConverterCore` class is opened and closed in the subplugins.
So converter should pass the double pointer.

File: ext/nnstreamer/tensor_converter/tensor_converter_python3.cc [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+
+    py_close (priv_data);
+  }
+
+  /* init null */
+  *priv_data = NULL;
+
+  core = new PYConverterCore (path);
+  if (core == NULL) {
+    ml_loge ("Failed to allocate memory for converter subplugin: Python\n");
```
[the remark this answers is not in the document]
Comment (gichan-jang) [reply ext/nnstreamer/tensor_converter/tensor_converter_python3.cc]: Updated. Thanks.

File: gst/nnstreamer/include/nnstreamer_plugin_api_converter.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (myungjoo) [reply gst/nnstreamer/include/nnstreamer_plugin_api_converter.h:83]: Then, use a single pointer for convert callback.

[the remark this answers is not in the document]
Comment (gichan-jang) [reply gst/nnstreamer/include/nnstreamer_plugin_api_converter.h:83]: Changed to single pointer for convert callback. Thanks!

**Review by myungjoo — APPROVED** [review]
LGTM

**Pull request merged**